### PR TITLE
Add ipv6 support

### PR DESCRIPTION
**Pull request by prasadlvi**

### CODE REVIEW

File: src/main/kotlin/com/zaxxer/ping/IcmpPinger.kt [hunk 1]
```diff
@@ -246,7 +252,11 @@ class IcmpPinger(private val responseHandler:PingResponseHandler) {
          val pingTarget = pendingPings.peek()
          if (sendIcmp(pingTarget, fd)) {
             pendingPings.take()
-            waitingTarget4Map[pingTarget.sequence] = pingTarget
+            if(pingTarget.isIPv4) {
```
Comment (brettwooldridge) [inline src/main/kotlin/com/zaxxer/ping/IcmpPinger.kt]: Whitespace after ``if`` please.  ``if`` is not a function or method. 😄

File: src/main/kotlin/com/zaxxer/ping/IcmpPinger.kt [hunk 2]
```diff
@@ -294,18 +304,29 @@ class IcmpPinger(private val responseHandler:PingResponseHandler) {
...
-      icmp.icmp_type.set(ICMP_ECHO)
-      icmp.icmp_code.set(0)
-      icmp.icmp_cksum.set(0)
-
-      // In BSD we are responsible for the entire payload, including checksum.  Linux mucks with the payload (replacing
-      // the identity field, and therefore recalculates the checksum (so don't waste our time doing it here).
-      if (isBSD) {
-         val cksum = icmpCksum(outpacketPointer, SEND_PACKET_SIZE)
-         icmp.icmp_cksum.set(htons(cksum.toShort()))
+      if(pingTarget.isIPv4) {
```
Comment (brettwooldridge) [inline src/main/kotlin/com/zaxxer/ping/IcmpPinger.kt]: Whitespace after if please. if is not a function or method. 😄

File: src/main/kotlin/com/zaxxer/ping/IcmpPinger.kt [hunk 3]
```diff
@@ -333,11 +354,18 @@ class IcmpPinger(private val responseHandler:PingResponseHandler) {
          cc -= headerLen
 
          icmp.useMemory(socketBufferPointer.slice(headerLen.toLong()))
-         if (icmp.icmp_type.get() != ICMP_ECHOREPLY) {
+         icmp6.useMemory(socketBufferPointer)
+
+         if (icmp.icmp_type.get() != ICMP_ECHOREPLY && icmp6.icmp6_type.get() != ICMPV6_ECHO_REPLY) {
```
Comment (brettwooldridge) [inline src/main/kotlin/com/zaxxer/ping/IcmpPinger.kt]: Isn't this dangerous?  Are the ``icmp.icmp_type`` and ``icmp6.icmp6_type`` fields at the same location in the reply packet?  Is it possible for the ``icmp.icmp_type`` memory location to have a value of ``ICMP_ECHOREPLY`` just by accident in the case of an IPv6 reply?

File: src/main/kotlin/com/zaxxer/ping/impl/IcmpNative.kt [hunk 4]
```diff
@@ -67,16 +68,25 @@ class BSDSockAddr4(address:Inet4Address) : SockAddr() {
...
+   @field:JvmField val sin6_port = Unsigned16()
+   @field:JvmField val sin6_flowinfo = Unsigned32()
+   @field:JvmField val sin6_addr:Array<out Unsigned8> = Array(16, {Unsigned8()})
+   @field:JvmField val sin6_scope_id = Unsigned32()
 
    init {
-      sin_family.set(PF_INET6)
+      sin6_family.set(PF_INET6)
+      val bytes = address.address
+      for(index in 0..15) {
```
Comment (brettwooldridge) [inline src/main/kotlin/com/zaxxer/ping/impl/IcmpNative.kt]: Whitespace after ``for``.

File: src/main/kotlin/com/zaxxer/ping/impl/IcmpNative.kt [hunk 5]
```diff
@@ -67,16 +68,25 @@ class BSDSockAddr4(address:Inet4Address) : SockAddr() {
...
+         val value = bytes[index].toInt() and 0xff
+         sin6_addr[index].set(value)
+      }
+
+      if(address.isLinkLocalAddress) {
```
Comment (brettwooldridge) [inline src/main/kotlin/com/zaxxer/ping/impl/IcmpNative.kt]: Whitespace after ``if``.

File: src/main/kotlin/com/zaxxer/ping/IcmpPinger.kt [hunk 6]
```diff
@@ -333,11 +354,18 @@ class IcmpPinger(private val responseHandler:PingResponseHandler) {
...
             LOGGER.fine({"   ^ Opps, not our response."})
          }
          else {
-            val seq = ntohs(icmp.icmp_hun.ih_idseq.icd_seq.shortValue())
+            val seq = if(icmp.icmp_type.get() == ICMP_ECHOREPLY) {
```
Comment (brettwooldridge) [inline src/main/kotlin/com/zaxxer/ping/IcmpPinger.kt]: Whitespace after ``if``.

File: src/main/kotlin/com/zaxxer/ping/impl/IcmpNative.kt [hunk 7]
```diff
@@ -94,12 +104,26 @@ class LinuxSockAddr4(address:Inet4Address) : SockAddr() {
...
+   @field:JvmField val sin6_family = Unsigned16()
+   @field:JvmField val sin6_port = Unsigned16()
+   @field:JvmField val sin6_flowinfo = Unsigned32()
+   @field:JvmField val sin6_addr:Array<out Unsigned8> = Array(16, {Unsigned8()})
+   @field:JvmField val sin6_scope_id = Unsigned32()
+
+   init {
+      sin6_family.set(PF_INET6)
+      val bytes = address.address
+      for(index in 0..15) {
```
Comment (brettwooldridge) [inline src/main/kotlin/com/zaxxer/ping/impl/IcmpNative.kt]: Whitespace after ``for``.

File: src/main/kotlin/com/zaxxer/ping/impl/IcmpNative.kt [hunk 8]
```diff
@@ -94,12 +104,26 @@ class LinuxSockAddr4(address:Inet4Address) : SockAddr() {
...
+         val value = bytes[index].toInt() and 0xff
+         sin6_addr[index].set(value)
+      }
+
+      if(address.isLinkLocalAddress) {
```
Comment (brettwooldridge) [inline src/main/kotlin/com/zaxxer/ping/impl/IcmpNative.kt]: Whitespace after ``if``.

File: src/main/kotlin/com/zaxxer/ping/IcmpPinger.kt (hunk 3 above)
Comment (prasadlvi) [reply src/main/kotlin/com/zaxxer/ping/IcmpPinger.kt]: icmp.icmp_type and icmp6.icmp6_type are in the same location of the reply packet.

File: src/main/kotlin/com/zaxxer/ping/IcmpPinger.kt (hunk 1 above)
Comment (prasadlvi) [reply src/main/kotlin/com/zaxxer/ping/IcmpPinger.kt]: Sorry for formatting errors.

File: src/main/kotlin/com/zaxxer/ping/IcmpPinger.kt (hunk 3 above)
Comment (prasadlvi) [reply src/main/kotlin/com/zaxxer/ping/IcmpPinger.kt]: icmp.icmp_type

```
class Icmp:Struct(runtime) {
   // u_char	icmp_type;		/* type of message, see below */
   // u_char	icmp_code;		/* type sub code */
   // u_short	icmpCksum;		/* ones complement cksum of struct */
   val icmp_type = Unsigned8()
   val icmp_code = Unsigned8()
   val icmp_cksum = Unsigned16()
   // union {
   val icmp_hun:Hun = inner(Hun())
   // } icmp_hun;
   // union {
   val icmp_dun:Dun = inner(Dun())
   // } icmp_dun
}
```

icmp6.icmp6_type

```
class Icmp6:Struct(runtime) {
   // u_char	icmp_type;		/* type of message, see below */
   // u_char	icmp_code;		/* type sub code */
   // u_short	icmpCksum;		/* ones complement cksum of struct */
   val icmp6_type = Unsigned8()
   val icmp6_code = Unsigned8()
   val icmp6_cksum = Unsigned16()
   // union {
   val icmp6_dataun:Icmp6UnData = inner(Icmp6UnData())
   // } icmp6_dataun;
}
```

**Comment by prasadlvi** [comment]
I updated the pull request with suggested changes.

File: src/main/kotlin/com/zaxxer/ping/IcmpPinger.kt [hunk 9]
```diff
@@ -333,11 +354,19 @@ class IcmpPinger(private val responseHandler:PingResponseHandler) {
          cc -= headerLen
 
          icmp.useMemory(socketBufferPointer.slice(headerLen.toLong()))
-         if (icmp.icmp_type.get() != ICMP_ECHOREPLY) {
-            LOGGER.fine({"   ^ Opps, not our response."})
-         }
-         else {
-            val seq = ntohs(icmp.icmp_hun.ih_idseq.icd_seq.shortValue())
+         icmp6.useMemory(socketBufferPointer)
```
Comment (brettwooldridge) [inline src/main/kotlin/com/zaxxer/ping/IcmpPinger.kt]: Is it true that we do not need to "slice()" the ``socketBufferPointer`` like we do for ``icmp`` (line 356)?

**Pull request merged**